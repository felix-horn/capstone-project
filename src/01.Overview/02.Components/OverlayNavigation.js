import { useHistory } from 'react-router-dom'
import PropTypes from 'prop-types'
import { v4 as uuid } from 'uuid'
import styled from 'styled-components/macro'
import ButtonLabeledCircle from '../01.UI-Elements/ButtonLabeledCircle'
import AddShopIcon from '@material-ui/icons/PostAdd'
import ScannerIcon from '@material-ui/icons/CropFree'

OverlayNavigation.propTypes = {
  onClick: PropTypes.func.isRequired,
  addShop: PropTypes.func.isRequired,
  className: PropTypes.string,
}

export default function OverlayNavigation({ onClick, addShop, className }) {
  const history = useHistory()
  const newShopId = uuid()

  return (
    <>
      <OpaqueBackground onClick={onClick} className={className} />
      <ButtonAddShop
        title="Neues Geschäft hinzufügen"
        onClick={handleAddShopClick}
        variant="small"
      >
        <AddShopIcon />
      </ButtonAddShop>
      <ButtonScanner
        title="Barcode scannen"
        onClick={handleScanBarcodeClick}
        variant="primary"
      >
        <ScannerIconAnimated />
      </ButtonScanner>
    </>
  )
  function handleAddShopClick() {
    history.push({
      pathname: '/shop',
      state: { shopId: newShopId },
    })
    addShop(newShopId)
  }
  function handleScanBarcodeClick() {
    history.push({
      pathname: '/scanner',
      state: { useCase: 'uncheckItem' },
    })
  }
}

const OpaqueBackground = styled.div`
  height: 100%;
  width: 100%;
  background-color: var(--white);
  opacity: 0.8;
`
const ButtonAddShop = styled(ButtonLabeledCircle)`
  position: absolute;
  z-index: var(--z-index-item-on-overlay);
  bottom: calc(120px + 10px + 5px);
  right: calc(30px + 10px + 5px);

  animation: 0.2s ease-in-out 0s 1 slideUp;

  @keyframes slideUp {
    0% {
      transform: translateY(20px);
      opacity: 0;
    }
    100% {
      transform: translateY(0px);
      opacity: 1;
    }
  }
`
const ButtonScanner = styled(ButtonLabeledCircle)`
  position: absolute;
  z-index: var(--z-index-item-on-overlay);
  bottom: calc(60px + 10px);
  right: calc(30px + 10px);
`
const ScannerIconAnimated = styled(ScannerIcon)`
  animation: 0.2s ease-in-out 0s 1 rotate;

  @keyframes rotate {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(90deg);
    }
  }
`
