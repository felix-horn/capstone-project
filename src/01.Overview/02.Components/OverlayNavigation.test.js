/* eslint-disable no-undef */
import { render, fireEvent } from '@testing-library/react'
import user from '@testing-library/user-event'
import OverlayNavigation from './OverlayNavigation'

describe('OverlayNavigation', () => {
  it('renders correctly', () => {
    const onBackgroundClickMock = jest.fn()
    const addShopMock = jest.fn()
    const { container } = render(
      <OverlayNavigation
        onBackgroundClick={onBackgroundClickMock}
        addShop={addShopMock}
      />
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('calls function onBackgroundClick', () => {
    const onBackgroundClickMock = jest.fn()
    const addShopMock = jest.fn()
    const { getByTestId } = render(
      <OverlayNavigation
        onBackgroundClick={onBackgroundClickMock}
        addShop={addShopMock}
      />
    )
    const background = getByTestId('opaque-background')
    user.click(background)
    expect(onBackgroundClickMock).toHaveBeenCalledTimes(1)
  })

  it('calls function addShop', () => {
    const onBackgroundClickMock = jest.fn()
    const addShopMock = jest.fn()
    const { getByTestId, getByText } = renderWithRouter(
      <OverlayNavigation
        onBackgroundClick={onBackgroundClickMock}
        addShop={addShopMock}
      />
    )
    // why cant the button be gotten by its own test ID?
    // probably because it already got one in general
    // but I wanna get that specific one by an ID
    // const button = getByTestId('button-add-shop')
    const button = getByText('Neues Geschäft hinzufügen')
    user.click(button)
    expect(addShopMock).toHaveBeenCalledTimes(1)
  })
})
