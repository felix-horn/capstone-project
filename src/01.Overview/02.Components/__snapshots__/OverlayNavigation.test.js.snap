// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverlayNavigation renders correctly 1`] = `
.c0 {
  height: 100%;
  width: 100%;
  background-color: var(--white);
  opacity: 0.8;
}

<div
  class="c0"
  data-testid="opaque-background"
/>
`;
